src/planner/kinodynamic/KinodynamicTimer.cpp

```diff
@@ -0,0 +1,273 @@
+
+    // create a sequence of time steps from start time to end time by time step
+    aikido::common::StepSequence sequence(
+        timeStep, true, true, startTime, endTime);
+
+    const Eigen::VectorXd zeroPosition = Eigen::VectorXd::Zero(dimension);
+    auto currState = stateSpace->createState();
+    for (std::size_t i = 1; i < sequence.getLength(); i++)
+    {
+      double currT = sequence[i - 1];
+      double nextT = sequence[i];
+      double segmentDuration = nextT - currT;
+      double currTShift = currT - startTime;
+      double nextTShift = nextT - startTime;
+      Eigen::VectorXd currentPosition = trajectory.getPosition(currTShift);
+      Eigen::VectorXd nextPosition = trajectory.getPosition(nextTShift);
+      Eigen::VectorXd currentVelocity = trajectory.getVelocity(currTShift);
+      Eigen::VectorXd nextVelocity = trajectory.getVelocity(nextTShift);
+
+      CubicSplineProblem problem(
+          Eigen::Vector2d{0., segmentDuration}, 4, dimension);
+      problem.addConstantConstraint(0, 0, zeroPosition);
+      problem.addConstantConstraint(0, 1, currentVelocity);
+      problem.addConstantConstraint(1, 0, nextPosition - currentPosition);
+      problem.addConstantConstraint(1, 1, nextVelocity);
+      const auto solution = problem.fit();
+      const auto coefficients = solution.getCoefficients().front();
+
+      stateSpace->expMap(currentPosition, currState);
+      outputTrajectory->addSegment(coefficients, segmentDuration, currState);
+    }
+
+    return outputTrajectory;
+  }
+
+  // path is invalid. Retiming failed.
+  return nullptr;
+}
+
+//==============================================================================
+std::unique_ptr<aikido::trajectory::Spline> computeKinodynamicTiming(
+    const aikido::trajectory::Interpolated& inputTrajectory,
+    const Eigen::VectorXd& maxVelocity,
+    const Eigen::VectorXd& maxAcceleration,
+    double maxDeviation,
+    double timeStep)
+{
+  const auto stateSpace = inputTrajectory.getStateSpace();
+  const auto dimension = stateSpace->getDimension();
+
+  if (static_cast<std::size_t>(maxVelocity.size()) != dimension)
+    throw std::invalid_argument("Velocity limits have wrong dimension.");
+
+  if (static_cast<std::size_t>(maxAcceleration.size()) != dimension)
+    throw std::invalid_argument("Acceleration limits have wrong dimension.");
+
+  for (std::size_t i = 0; i < dimension; ++i)
+  {
+    if (maxVelocity[i] <= 0.)
+      throw std::invalid_argument("Velocity limits must be positive.");
+    if (!std::isfinite(maxVelocity[i]))
+      throw std::invalid_argument("Velocity limits must be finite.");
+
+    if (maxAcceleration[i] <= 0.)
+      throw std::invalid_argument("Acceleration limits must be positive.");
+    if (!std::isfinite(maxAcceleration[i]))
+      throw std::invalid_argument("Acceleration limits must be finite.");
+  }
+
+  double startTime = inputTrajectory.getStartTime();
+  // create waypoints from Interpolated path
```

Comment:
I think it would be nice to create `Eigen::VectorXd getWaypoints(trajectory::Interpolated traj)` (and for splines), then call the appropriate function from here.

Reply:
Done! Two new functions are added.